src/software/ai/navigator/obstacle/obstacle_factory.cpp

```diff
@@ -1,101 +1,193 @@
-                                 ROBOT_MAX_RADIUS_METERS);
-                obstacles.push_back(Obstacle(rectangle));
-                break;
-            case MotionConstraint::ENEMY_DEFENSE_AREA:
-                // We extend the enemy defense area back by several meters to prevent
-                // robots going around the back of the goal
-                rectangle = Rectangle(
-                    world.field().enemyDefenseArea().negXPosYCorner(),
-                    Point(10, world.field().enemyDefenseArea().negXNegYCorner().y()));
-                rectangle.expand(config->RobotObstacleInflationFactor()->value() *
-                                 ROBOT_MAX_RADIUS_METERS);
-                obstacles.push_back(Obstacle(rectangle));
-                break;
-            case MotionConstraint::INFLATED_ENEMY_DEFENSE_AREA:
-                rectangle = world.field().enemyDefenseArea();
-                rectangle.expand(config->RobotObstacleInflationFactor()->value() *
-                                     ROBOT_MAX_RADIUS_METERS +
-                                 0.3);  // 0.3 is by definition what inflated means
-                obstacles.push_back(Obstacle(rectangle));
-                break;
-            case MotionConstraint::CENTER_CIRCLE:
-                obstacles.push_back(Obstacle::createCircleObstacle(
-                    world.field().centerPoint(), world.field().centerCircleRadius(),
-                    config->RobotObstacleInflationFactor()->value()));
-                break;
-            case MotionConstraint::HALF_METER_AROUND_BALL:
-                obstacles.push_back(Obstacle::createCircleObstacle(
-                    world.ball().position(), 0.5,  // 0.5 represents half a metre radius
-                    config->RobotObstacleInflationFactor()->value()));
-                break;
-            case MotionConstraint::ENEMY_HALF:
-                rectangle = Rectangle({0, world.field().totalYLength() / 2},
-                                      world.field().enemyCornerNeg() -
-                                          Vector(0, world.field().boundaryYLength()));
-                rectangle.expand(config->RobotObstacleInflationFactor()->value() *
-                                 ROBOT_MAX_RADIUS_METERS);
-                obstacles.push_back(Obstacle(rectangle));
-                break;
-            case MotionConstraint::FRIENDLY_HALF:
-                rectangle = Rectangle({0, world.field().totalYLength() / 2},
-                                      world.field().friendlyCornerNeg() -
-                                          Vector(0, world.field().boundaryYLength()));
-                rectangle.expand(config->RobotObstacleInflationFactor()->value() *
-                                 ROBOT_MAX_RADIUS_METERS);
-                obstacles.push_back(Obstacle(rectangle));
-                break;
+            obstacles.push_back(createObstacle(*rectangle_ptr));
```

Comment:
Seems like you should use the `createObstacleFromRectangle` function here? Also `rectangle_ptr` is an optional, not a pointer so the name is somewhat misleading.